art/metrics/privacy/membership_leakage_metrics.py

```diff
@@ -0,0 +1,89 @@
+    """
+        Compute the pointwise differential training privacy metric for the given classifier and training set.
+        Taken from: https://arxiv.org/abs/1712.09136
+
+        :param target_estimator: The classifier to be analyzed.
+        :param extra_estimator: Another classifier of the same type as the target classifier, but not yet fit.
+        :param x: The training data of the classifier.
+        :param y: Target values (class labels) of `x`, one-hot-encoded of shape (nb_samples, nb_classes) or indices of
+                  shape (nb_samples,).
+        :param indexes: the subset of indexes of `x` to compute the PDTP metric on. If not supplied, PDTP will be
+                        computed for all samples in `x`.
+        :param num_iter: the number of iterations of PDTP computation to run for each sample. If not supplied,
+                         defaults to 10. The result is the average across iterations.
+        :return: an array containing the average PDTP value for each sample in the training set.
+    """
+    if ClassifierMixin not in type(target_estimator).__mro__ or ClassifierMixin not in type(extra_estimator).__mro__:
+        raise ValueError("PDTP metric only supports classifiers")
+    if target_estimator.input_shape[0] != x.shape[1]:
+        raise ValueError("Shape of x does not match input_shape of classifier")
+    y = check_and_transform_label_format(y, target_estimator.nb_classes)
+    if y.shape[0] != x.shape[0]:
+        raise ValueError("Number of rows in x and y do not match")
+
+    results = []
+
+    for i in range(num_iter):
+        iter_results = []
+        # get probabilities from original model
+        pred = target_estimator.predict(x)
+        if not is_probability(pred):
+            try:
+                pred = scipy.special.softmax(pred, axis=1)
+            except:
+                raise ValueError(
+                    "PDTP metric only supports classifiers that output logits or probabilities."
+                )
+        if not indexes:
+            indexes = range(x.shape[0])
+        for row in indexes:
+            # create new model without sample in training data
+            alt_x = np.delete(x, row, 0)
+            alt_y = np.delete(y, row, 0)
+            try:
+                extra_estimator.reset()
+            except NotImplementedError:
+                raise ValueError(
+                    "PDTP metric can only be applied to classifiers that implement the reset method."
+                )
+            extra_estimator.fit(alt_x, alt_y)
+            # get probabilities from new model
+            alt_pred = extra_estimator.predict(x)
+            if not is_probability(alt_pred):
+                alt_pred = scipy.special.softmax(alt_pred, axis=1)
+            # divide into 100 bins and return center of bin
+            pred_bin = (np.floor(pred * 100) / 100).round(decimals=2) + 0.005
+            pred_bin[pred_bin > 1] = 0.995
```

Comment:
It looks like these lines are independent of `row`, would it be possible to move these lines to before the for-loop in line 56?

Reply:
Could it be helpful to use `numpy.histogram` here?

Reply:
I don't think that numpy.histogram can help here because I need to know to which bin each sample belongs, but maybe numpy.digitize would work. I'll check.

Reply:
Ok.